core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala

```diff
@@ -99,13 +118,83 @@ private class FallbackConfigEntry[T] (
     key: String,
     doc: String,
     isPublic: Boolean,
-    private val fallback: ConfigEntry[T])
-    extends ConfigEntry[T](key, fallback.valueConverter, fallback.stringConverter, doc, isPublic) {
+    private[config] val fallback: ConfigEntry[T],
+    private val expandVars: Boolean)
+    extends ConfigEntry[T](
+        key,
+        fallback.valueConverter,
+        fallback.stringConverter,
+        doc,
+        isPublic,
+        expandVars) {
 
   override def defaultValueString: String = s"<value of ${fallback.key}>"
 
-  override def readFrom(conf: SparkConf): T = {
-    conf.getOption(key).map(valueConverter).getOrElse(fallback.readFrom(conf))
+  override def readFrom(conf: JMap[String, String], getenv: String => String): T = {
+    Option(conf.get(key)).map(valueConverter).getOrElse(fallback.readFrom(conf, getenv))
+  }
+
+}
+
+private object ConfigEntry {
+
+  private val knownConfigs = new java.util.concurrent.ConcurrentHashMap[String, ConfigEntry[_]]()
+
+  private val REF_RE = "\\$\\{(?:(\\w+?):)?(\\S+?)\\}".r.pattern
+
+  def registerEntry(entry: ConfigEntry[_]): Unit = {
+    val existing = knownConfigs.putIfAbsent(entry.key, entry)
+    require(existing == null, s"Config entry ${entry.key} already registered!")
+  }
+
+  def findEntry(key: String): ConfigEntry[_] = knownConfigs.get(key)
+
+  /**
+   * Expand the `value` according to the rules explained in `ConfigBuilder.withVariableExpansion`.
+   */
+  def expand(
+      value: String,
+      conf: JMap[String, String],
+      getenv: String => String,
+      usedRefs: Set[String]): String = {
+    val matcher = REF_RE.matcher(value)
+    val result = new StringBuilder()
+    var end = 0
+
+    while (end < value.length() && matcher.find(end)) {
```

Comment:
I think you can use [Regex#replaceAllIn](http://daily-scala.blogspot.com/2010/02/regex-replaceallin.html) here instead of a loop.